### ISSUE TITLE
Exception: A DOM Element reference is required

### ISSUE THREAD

**Issue by OliverJAsh** [WICG/focus-ring#56]
I just put this into production on unsplash.com and noticed that we're getting thousands of client-side exceptions reported due to `A DOM Element reference is required`. The offending line is here https://github.com/WICG/focus-ring/blob/0e9f086bbc59ca84a2a932e89df0b71bce18b643/dist/focus-ring.js#L25

I haven't been able to reproduce this locally. Do you have any ideas why this might be happening? Have you seen it before?

Thanks!

**Comment by OliverJAsh** [comment]
Almost every occurrence of the error seems to be IE11. 🤔 I just tried accessing the website in Browserstack and I couldn't reproduce.

**Comment by OliverJAsh** [comment]
Here are stack traces:

```
Error: A DOM Element reference is required
  at e(./node_modules/wicg-focus-ring/dist/focus-ring.js:25:0)
  at index(./node_modules/wicg-focus-ring/dist/focus-ring.js:15:0)
  at u(./node_modules/wicg-focus-ring/dist/focus-ring.js:302:0)
  at n(./node_modules/raven-js/src/raven.js:321:0)
```

and

```
Error: A DOM Element reference is required
  at e(./node_modules/wicg-focus-ring/dist/focus-ring.js:25:0)
  at index(./node_modules/wicg-focus-ring/dist/focus-ring.js:15:0)
  at addFocusRingClass(./node_modules/wicg-focus-ring/dist/focus-ring.js:225:0)
  at s(./node_modules/wicg-focus-ring/dist/focus-ring.js:292:0)
  at n(./node_modules/raven-js/src/raven.js:321:0)
```

It looks like `document.activeElement` is failing to pass the condition `!el || el.nodeType !== 1`.

https://github.com/WICG/focus-ring/blob/0e9f086bbc59ca84a2a932e89df0b71bce18b643/dist/focus-ring.js#L24

**Comment by cycomachead** [comment]
We are seeing this in our error logs too. I haven't updated to v2...yet, though.

**Comment by OliverJAsh** [comment]
We are using the latest release, v2.0.1

**Comment by robdodson** [comment]
Sorry for the delay, both Alice and I are traveling at the moment.  I'll see if I can look into this a bit tomorrow.

**Comment by robdodson** [comment]
Looking into this a bit more, I wonder if this line is causing the issue:

https://github.com/WICG/focus-ring/blob/0e9f086bbc59ca84a2a932e89df0b71bce18b643/dist/focus-ring.js#L302

If there's no activeElement it may fall back to `document` which doesn't have a nodeType of 1 https://github.com/WICG/focus-ring/blob/0e9f086bbc59ca84a2a932e89df0b71bce18b643/dist/focus-ring.js#L24

I'll keep digging just wanted to give a little status update :D

**Comment by robdodson** [comment]
yeah double checking your stack traces, they're both emanating from these code blocks:

https://github.com/WICG/focus-ring/blob/0e9f086bbc59ca84a2a932e89df0b71bce18b643/dist/focus-ring.js#L290-L308

**Comment by robdodson** [comment]
I'm also having a hard time reproducing this.

@OliverJAsh, @cycomachead would y'all mind adding a snippet to your version of `dist/focus-ring.js` to help us debug?

This would replace [the ClassList function at line 23](https://github.com/WICG/focus-ring/blob/0e9f086bbc59ca84a2a932e89df0b71bce18b643/dist/focus-ring.js#L23)

```js
/**
 * Initialize a new ClassList for the given element
 *
 * @param {Element} el DOM Element
 */
function ClassList(el) {
  if (!el) {
    throw new Error('The element passed to ClassList was undefined');
  }

  if (el.nodeType !== 1) {
    throw new Error('Element ' + el.nodeName + ' is wrong nodeType: ' + el.nodeType);
  }

  this.el = el;
  this.classList = el.classList;
}
```

**Comment by robdodson** [comment]
Friendly ping on this. If either of you are able to include that snippet in your site so we can track down the source of the error that would be super helpful :)

**Comment by cycomachead** [comment]
Hey, sorry for the slow response. Unfortunately, I just haven't gotten a chance to push the fix to production because web pack doesn't like me (or something). I'll let you know when I get some data! :)

**Comment by OliverJAsh** [comment]
@robdodson Just got back from holiday, adding it now. I'll let you know.

**Comment by rodneyrehm** [comment]
When removing the activeElement from DOM it's possible IE11 is in state `document.activeElement === null`. When the browser window is unfocused (shift focus to another application, or F12 Tools) and refocused, the handlers `onWindowBlur()` and `onWindowFocus()` cause the ClassList to be constructed with `null` as the DOM Element to operate on, hence the `A DOM Element reference is required` error.

The simplest mitigation is to add the guard `if (!document.activeElement) return;` to `onWindowBlur()` and `onWindowFocus()`.

See https://github.com/medialize/ally.js/blob/master/src/get/active-element.js#L12-L32 for a few more edge cases regarding `document.activeElement`

**Comment by OliverJAsh** [comment]
@robdodson Given the above comment by @rodneyrehm, do you still need me to add the debug script?

**Comment by robdodson** [comment]
I think this should be fixed now. I just published 2.0.2 if you want to give it a shot :)

Thanks @rodneyrehm for pointing us in the right direction!

**Comment by rodneyrehm** [comment]
@robdodson since upgrading to 3.0.0 I'm getting this error again.

![screen shot 2017-11-21 at 18 20 01](https://user-images.githubusercontent.com/186837/33086780-fdbb909c-cee8-11e7-8624-2b714f5f5ded.png)

**Comment by robdodson** [comment]
oh shoot sorry! I backed out some changes that had broken our tests and must have missed this fix. I'll add it back in. I'll also add a reminder to see if we can get our test suite to reproduce it.

**Comment by robdodson** [comment]
Just published 3.0.2 which should contain this fix. Sorry about that! Please let me know if you hit any other issues.
